onnxruntime/core/providers/cuda/reduction/reduction_ops.cc

```diff
@@ -273,6 +301,83 @@ Status ReduceKernel<allow_multi_axes>::ComputeImpl(OpKernelContext* ctx, cudnnRe
+                                       &Y,
+                                       input_count,
+                                       output_count,
+                                       output_dims,
+                                       input_dims_cudnn,
+                                       output_dims_cudnn));
+
+  // special case when there is a dim value of 0 in the shape.
+  if (input_count == 0) {
+    assert(Y->Shape().Size() == 0);
+    return Status::OK();
+  }
+
+  // cudnnReduceTensor for ReduceSum has issue if input and output has same size, we just need to copy the data for this case
+  if (input_count == output_count) {
+    if (Y->template MutableData<int32_t>() != X->template Data<int32_t>()) {
+      CUDA_RETURN_IF_ERROR(cudaMemcpyAsync(Y->template MutableData<int32_t>(), X->template Data<int32_t>(), input_count * sizeof(int32_t), cudaMemcpyDeviceToDevice));
+    }
+    return Status::OK();
+  }
+
+  size_t indices_bytes = 0;
+  size_t workspace_bytes = 0;
+  CudnnTensor input_tensor;
+  CudnnTensor output_tensor;
+  CudnnReduceDescriptor reduce_desc;
+
+  cudnnDataType_t cudnn_type_X = CUDNN_DATA_FLOAT;
+  IAllocatorUniquePtr<float> temp_X = GetScratchBuffer<float>(input_count);
+  Impl_Cast<CudaT, float>(reinterpret_cast<const CudaT*>(X->template Data<int32_t>()), temp_X.get(), X->Shape().Size());
+
+  ORT_RETURN_IF_ERROR(reduce_desc.Set(cudnnReduceOp, cudnn_type_X, CUDNN_REDUCE_TENSOR_FLATTENED_INDICES));
+  ORT_RETURN_IF_ERROR(input_tensor.Set(input_dims_cudnn, cudnn_type_X));
+  ORT_RETURN_IF_ERROR(output_tensor.Set(output_dims_cudnn, cudnn_type_X));
+  CUDNN_RETURN_IF_ERROR(cudnnGetReductionIndicesSize(CudnnHandle(), reduce_desc, input_tensor, output_tensor, &indices_bytes));
+  CUDNN_RETURN_IF_ERROR(cudnnGetReductionWorkspaceSize(CudnnHandle(), reduce_desc, input_tensor, output_tensor, &workspace_bytes));
+  IAllocatorUniquePtr<uint32_t> indices_cuda = GetScratchBuffer<uint32_t>(indices_bytes);
+  IAllocatorUniquePtr<CudaT> workspace_cuda = GetScratchBuffer<CudaT>(workspace_bytes);
+
+  const auto one = Consts<float>::One;
+  const auto zero = Consts<float>::Zero;
+  auto temp_Y = GetScratchBuffer<float>(output_count);
+  CUDNN_RETURN_IF_ERROR(cudnnReduceTensor(CudnnHandle(),
+                                          reduce_desc,
+                                          indices_cuda.get(),
+                                          indices_bytes,
+                                          workspace_cuda.get(),
+                                          workspace_bytes,
+                                          &one,
+                                          input_tensor,
+                                          temp_X.get(),
+                                          &zero,
+                                          output_tensor,
+                                          temp_Y.get()));
+
+  Impl_Cast<float, int32_t>(temp_Y.get(), Y->template MutableData<int32_t>(), output_count);
```

Comment:
Cast from float to int32 might have problem. For some float like *.9999997, what will be expected result of cast?